api/src/main/java/org/pmiops/workbench/elasticsearch/ElasticFilters.java

```diff
@@ -0,0 +1,254 @@
+   * an approximation, depending on the contents of the request.
+   */
+  public static class ElasticFilterResponse<T> {
+
+    private final T value;
+    private final boolean isApproximate;
+
+    public ElasticFilterResponse(T value, boolean isApproximate) {
+      this.value = value;
+      this.isApproximate = isApproximate;
+    }
+
+    public T value() {
+      return this.value;
+    }
+
+    public boolean isApproximate() {
+      return this.isApproximate;
+    }
+  }
+
+  /**
+   * Translates a Cohort Builder search request into an Elasticsearch filter. If the request
+   * parameters are not supported, this is indicated in the response and a best effort approximation
+   * of the target filter is made.
+   */
+  public static ElasticFilterResponse<QueryBuilder> fromCohortSearch(
+      CriteriaDao criteriaDao, SearchRequest req) {
+    ElasticFilters f = new ElasticFilters(criteriaDao);
+    QueryBuilder q = f.process(req);
+    return new ElasticFilterResponse<>(q, f.isApproximate);
+  }
+
+  private static Set<TreeType> STANDARD_TREES = ImmutableSet.of(
+      TreeType.SNOMED, TreeType.DRUG, TreeType.MEAS, TreeType.VISIT);
+
+  private final CriteriaDao criteriaDao;
+  private final Map<String, Set<Long>> criteriaToConceptChildren;
+  private boolean processed = false;
+  private boolean isApproximate = false;
+
+  private ElasticFilters(CriteriaDao criteriaDao) {
+    this.criteriaDao = criteriaDao;
+    this.criteriaToConceptChildren = Maps.newHashMap();
+  }
+
+  private QueryBuilder process(SearchRequest req) {
+    Preconditions.checkArgument(!processed);
+
+    for (SearchGroup sg : Iterables.concat(req.getIncludes(), req.getExcludes())) {
+      for (SearchGroupItem sgi : sg.getItems()) {
+        for (SearchParameter param : sgi.getSearchParameters()) {
+          if (!param.getGroup()) {
+            continue;
+          }
+          Preconditions.checkArgument(
+              !TreeType.SNOMED.toString().equals(param.getType()),
+              "SNOMED groups are poly-hierarchical; the SearchRequest does not encode enough " +
+                  "information to determine which criteria ID should be used");
+
+          // TODO: This two-step lookup is gross - dig into why we don't seem to have enough
+          // information available in the SearchRequest to more directly translate a criteria
+          // selection into a set of concept IDs. Also there is no uniqueness guarantee on codes
+          // nor do we seem to have documentation of what a "criteria code" is and where it comes
+          // from. Lastly, the code gets stuff into param.getValue() - this seems to be undocumented
+          // and unclear whether that field is polymorphic.
+          Criteria parent = criteriaDao.findCriteriaByTypeAndCodeAndGroup(param.getType(), param.getValue(), true);
+
+          // TODO: Try to batch this request for all search params.
+          Set<Long> children = criteriaDao.findCriteriaChildrenByTypeAndParentId(
```

Comment:
I think we can lookup children of parent in 3 different ways depending on search parameter type:
1) if ICD9/ICD10 and is_group = 1
`select concept_id
  from criteria
  where type = :type
  and subtype = :subtype
  and code REGEXP '^(:value)' 
  and is_selectable = 1
  and concept_id != ''`
2) if PPI and is_group = 1 and concept_id is null
`select concept_id
      from criteria
      where path like (
      select concat('%', id, '%') as path
      from criteria
      where subtype = :subtype
      and parent_id = 0)`
3) if Drug and is_group = 1
`select a.concept_id from
    criteria a
    join (select CONCAT( '%.', id, '%') as path
    from criteria
    where concept_id in (:conceptId)) b
    on a.path like b.path
    and is_group = 0
    and is_selectable = 1
    and type = 'DRUG'
    and subtype = 'ATC'`

Reply:
OK - I tried to generalize these cases out rather than tie them to specific tree types, see `buildCriteriaGroupLookup`. Let me know if this makes sense.

I decided to skip straight to minimal SQL calls rather than optimizing later, so the client code here is a bit more verbose. I think this is probably the effective minimal number of SQL calls we can make given our current SearchRequest requirements and SQL table schemas, without writing some crazy complex union query.